app/src/main/res/layout/fragment_pattern_editor.xml

```diff
@@ -51,4 +52,20 @@
                 android:layout_height="wrap_content"/>
         </TableRow>
     </TableLayout>
+    <com.google.android.material.textfield.TextInputLayout
```

Comment:
In my opinion, this should be a button in the action bar that displays a dialog where you can enter the text for both sender and content. Having the validation individually for sender/body seems weird to me, and can be misleading if one of {body,sender} matches but not the other. My idea for a design would be approximately:

```
Enter test data

Sender: ____________________________________
Body:   ____________________________________

🗹 Your filter matches this message.

                                        Done
```
```
Enter test data

Sender: ____________________________________
Body:   ____________________________________

🗷 Your filter does NOT match this message.

                                        Done
```

If there is no pattern entered for the sender/body, the corresponding textbox should be disabled and show `(any value)` as placeholder text.
```
Enter test data

Sender: (any value)
Body:   ____________________________________

🗹 Your filter matches this message.

                                        Done
```

Reply:
In most cases, you only need to test the body pattern, because on the one hand, the matching rules of the sender are relatively simple. On the other hand, the SMS does not block more situations is the problem of the body pattern, so I think we only need to display different prompts on different tabs instead of putting them together.